// simple-uniswap-sdk/dist/cjs/factories/router/models/route-methods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_ROUTER_METHOD = void 0;
exports.DEFAULT_ROUTER_METHOD = {
    WETH: 'WETH',
    addLiquidity: 'addLiquidity',
    addLiquidityETH: 'addLiquidityETH',
    factory: 'factory',
    getAmountIn: 'getAmountIn',
    getAmountOut: 'getAmountOut',
    getAmountsIn: 'getAmountsIn',
    getAmountsOut: 'getAmountsOut',
    quote: 'quote',
    removeLiquidity: 'removeLiquidity',
    removeLiquidityETH: 'removeLiquidityETH',
    removeLiquidityETHSupportingFeeOnTransferTokens: 'removeLiquidityETHSupportingFeeOnTransferTokens',
    removeLiquidityETHWithPermit: 'removeLiquidityETHWithPermit',
    removeLiquidityETHWithPermitSupportingFeeOnTransferTokens: 'removeLiquidityETHWithPermitSupportingFeeOnTransferTokens',
    removeLiquidityWithPermit: 'removeLiquidityWithPermit',
    swapETHForExactTokens: 'swapETHForExactTokens',
    swapExactETHForTokens: 'swapExactETHForTokens',
    swapExactETHForTokensSupportingFeeOnTransferTokens: 'swapExactETHForTokensSupportingFeeOnTransferTokens',
    swapExactTokensForETH: 'swapExactTokensForETH',
    swapExactTokensForETHSupportingFeeOnTransferTokens: 'swapExactTokensForETHSupportingFeeOnTransferTokens',
    swapExactTokensForTokens: 'swapExactTokensForTokens',
    swapExactTokensForTokensSupportingFeeOnTransferTokens: 'swapExactTokensForTokensSupportingFeeOnTransferTokens',
    swapTokensForExactETH: 'swapTokensForExactETH',
    swapTokensForExactTokens: 'swapTokensForExactTokens',
};
